internal/wclayer/converttobaselayer.go

```diff
@@ -0,0 +1,164 @@
+package wclayer
+
+import (
+	"context"
+	"os"
+	"path/filepath"
+	"strings"
+	"syscall"
+
+	"github.com/Microsoft/hcsshim/internal/hcserror"
+	"github.com/Microsoft/hcsshim/internal/oc"
+	"github.com/Microsoft/hcsshim/internal/safefile"
+	"github.com/Microsoft/hcsshim/internal/winapi"
+	"github.com/pkg/errors"
+	"go.opencensus.io/trace"
+)
+
+var hiveNames = []string{"DEFAULT", "SAM", "SECURITY", "SOFTWARE", "SYSTEM"}
+
+// Ensure the given file exists as an ordinary file, and create a zero-length file if not.
+func ensureFile(path string, root *os.File) error {
+	stat, err := safefile.LstatRelative(path, root)
+	if err != nil && os.IsNotExist(err) {
+		// Do I need to restore directory-modifcation times on the rest of the path?
+		// That doesn't seem necessary since the directory _is_ being modified now.
+		// See `func (w *baseLayerWriter) Close()` for context
```

Comment:
@kevpar or @ambarve would probably be able to better answer this. Based on my understanding, you will want to restore timestamps at some point in this PR since you're creating new files, directories, and adding new hardlinks in various places. Even if you're just taking an arbitrary directory and attempting to turn it into a base layer, there may still be other files there that rely on the original timestamps.

Reply:
I'd definitely be interested to hear more about this. At this point, I'm not attempting to restore anything else about the state of the system, e.g. any newly-created files are left there after this operation completes, so I think it would be weird if the directory modification times were _not_ updated in the process.

Reply:
Hmm, this is a good question. I think we would be okay with not restoring the timestamps in this case because this probably isn't an actual bootable container image if we are creating empty hive files. @katiewasnothere what other files in a newly created layer do you think would be dependent on these timestamps?

Reply:
That's a good point. Plus this appears to only be used to create hive files if they don't exist. If we never intend for these images to be bootable, then we definitely wouldn't have a problem. But if we ever extended this to ensure other files are present to make a bootable image then we might want to come back to this.

For reference: this is the original issue that prompted the timestamp reverting in base layer writer https://github.com/moby/moby/issues/41066

Reply:
Yeah, this code path is only hit if the hive files didn't already exist. When taking an existing bootable layer, e.g., nanoserver, importing it and the exporting it again, you end up with a near-identical tarball, except hardlink ordering, and the timestamps on some of the things _outside_ the _Files/_ and _UtilityVM/_ directories, plus the files in UtilityVM referenced by

> ❔ It's not clear if the exported tarball should invert the backup step, and redirect the BCD and BCD.LOG to use the bcd.bak and bcd.log.bak (etc) files that were saved during import.

in the PR description, i.e. [`mutatedFiles`](https://github.com/microsoft/hcsshim/blob/ae4dec96de0b303db68e8ea76bc5d60de827184e/pkg/ociwclayer/import.go#L20-L29) which are backed up but never restored.

That difference itself, on examination, is not a problem for this code, as it's happening at the ociwclayer level, and the only changes this code makes at that level is support for hardlinks.

Reply:
As a structural response...

Timestamp reverting makes sense for the base layer writer (and reader, if it triggered timestamp modification that we cared about) because the output should be deterministic, and tarball/disk should match.

However, this code-path is not base layer reader/writer, but mutation of files that will be _fed_ to the base layer reader. It would be _correct_ for this code to trigger something like https://github.com/moby/moby/issues/41066 because it _does_ modify something, and so the 'change detection' that caused the database rebuild is doing its job as-intended.

In fact, what we're doing in `ensureBaseLayer` is making it easier for callers to meet the requirements for generating a valid parentless tarball, codifying the pre-reqs to remove the need for all users to maintain similar sets of code, e.g., https://github.com/buildpacks/imgutil/blob/main/layer/windows_baselayer.go and https://github.com/containerd/continuity/blob/master/fs/fstest/file_windows.go.

It also provides the location to _change_ those pre-reqs if we find they have changed, as I fear is the case for some version between RS5 and 20H2.

Reply:
So my current feeling is that I can just drop this comment next time I rebase, once #947 lands.